### ISSUE TITLE
Data race when SCCACHE_DIR is full and cleans up.

### ISSUE THREAD

**Issue by TriplEight** [mozilla/sccache#407]
We're building using GitLab CI with docker-runners. `sccache` is installed in docker images we're using for building.
It doesn't make sense using shared cache for us, so we use local setup with just `SCCACHE_DIR`, `RUSTC_WRAPPER=sccache` and `SCCACHE_CACHE_SIZE`.

We are experiencing random failures of build jobs every time `SCCACHE_CACHE_SIZE` is topped.
Usually a failed job succeeds after the relaunch. But the failures are pretty frequent and we have to clean the cache dir manually.

Logs look like that:
```
   Compiling parity-rpc-client v1.4.0 (/builds/parity/parity-ethereum/cli-signer/rpc-client)
error: failed to execute compile
caused by: error reading compile response from server
caused by: Failed to read response header
caused by: failed to fill whole buffer
error: Could not compile `parity-rpc-client`.
```

I'm able to reproduce the problem locally, without GitLab CI, just having topped cache dir and running several compilations at once.

**Comment by TriplEight** [comment]
I've managed to find the useful piece of log with `RUST_LOG=sccache`:
```
thread '<unnamed>' panicked at 'Error removing file from cache: `"/ci-cache/parity-ethereum/sccache/3/e/3eed5896ab29512e79d4290e12fb5edc6e70216e9f6e92cfea1accdde548e1a72b84744f699413a757565a7a301ca4c721ba02eb77b252abf18d09b580b7a593"`: No such file or directory (os error 2)', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/lru-disk-cache-0.2.0/src/lib.rs:183:62
```

**Comment by TriplEight** [comment]
I dare to suggest that this is probably due to the data race condition when this file was already deleted by a parallel build process that reads the cache from the same directory `/ci-cache/parity-ethereum/sccache`.

**Comment by TriplEight** [comment]
Bigger cache size did not help.
Once it was filled, the same random errors with `error: failed to execute compile` as here:
https://gitlab.parity.io/parity/substrate/-/jobs/147461 started appearing again.

**Comment by TriplEight** [comment]
We are still experiencing this problem, meanwhile removing the cache when it's filled.

**Comment by TriplEight** [comment]
@luser please suggest if I'm doing something wrong.
TLDR: Multiple docker containers with `sccache` installed are calling one docker volume on the CI machine. 
And when the `SCCACHE_CACHE_SIZE` is capped, these multiple instances of `sccache` are trying to remove some old cache. The race condition happens when they target to remove the same piece, but one did it first.

**Comment by luser** [comment]
Using the filesystem cache as a shared cache among multiple concurrent sccache servers is not a supported use case. I'd suggest using one of the other existing storage backends. If you're running your builds in Docker it shouldn't be hard to run something like Redis in a container with a persistent volume so you can point your build containers at it for caching.

**Comment by TriplEight** [comment]
Thanks! So Redis will handle this situation with multiple concurrent sccache servers?
